GoogleTestAdapter/DiaResolver/PeParser.cs

```diff
@@ -104,17 +104,26 @@ private static class NativeMethods
             // ReSharper disable once InconsistentNaming
             public const ushort IMAGE_DIRECTORY_ENTRY_IMPORT = 1;
 
-            [DllImport("imageHlp.dll", CallingConvention = CallingConvention.Winapi, CharSet = CharSet.Ansi, BestFitMapping = false, ThrowOnUnmappableChar = true)]
+            [DllImport("imageHlp.dll", CallingConvention = CallingConvention.Winapi, CharSet = CharSet.Ansi, BestFitMapping = false, ThrowOnUnmappableChar = true, SetLastError = true)]
             public static extern bool MapAndLoad(string imageName, string dllPath, LOADED_IMAGE* loadedImage, bool dotDll, bool readOnly);
 
-            [DllImport("imageHlp.dll", CallingConvention = CallingConvention.Winapi)]
+            [DllImport("imageHlp.dll", CallingConvention = CallingConvention.Winapi, SetLastError = true)]
             public static extern bool UnMapAndLoad(ref LOADED_IMAGE loadedImage);
 
-            [DllImport("dbghelp.dll", CallingConvention = CallingConvention.Winapi)]
-            public static extern void* ImageDirectoryEntryToData(IntPtr pBase, byte mappedAsImage, ushort directoryEntry, uint* size);
+            [DllImport("dbghelp.dll", EntryPoint = "ImageDirectoryEntryToData", CallingConvention = CallingConvention.Winapi, SetLastError = true)]
+            private static extern void* _ImageDirectoryEntryToData(IntPtr pBase, byte mappedAsImage, ushort directoryEntry, uint* size);
+            public static void* ImageDirectoryEntryToData(IntPtr pBase, byte mappedAsImage, ushort directoryEntry, uint* size)
+            {
+                // DbgHelp function ImageDirectoryEntryToData returns null in some cases without calling SetLastError. We thus set last error before calling the actual function to have a known value in these cases.
+                SetLastError(0);
```

Comment:
From what I understood from [here](https://social.msdn.microsoft.com/Forums/en-US/bad14dd4-0705-471b-80d3-42bd0d0ef44c/how-to-use-dllimport-setlasterror?forum=csharplanguage), the marshaller will take care of resetting the last error...

Reply:
I also thought that, but did a little test:
SetLastError(1234);
ImageDirectoryEntryToData( ... ); // on the EmptyExe
int error = Marshal.GetLastWin32Error();

(all the called DLL functions are declared with SetLastError = true)
I get 1234 in "error", thus clearly the marshaller does not reset last error when calling a function where SetLastError = true.

From my tests, Marshal.GetLastWin32Error() will reset the error to error 2 (file not found ?!), but I would not rely on that since it is not specified in the documentation, and anyway it is a weird choice of error number.

Reply:
Note that calling SetLastError before calling ImageDirectoryEntryToData is theoretically still not perfectly safe according to the doc of GetLastWin32Error, as a call to Windows API could be inserted between the two.  Thus, other than asking MS to correct the problem in ImageDirectoryEntryToData, if we want to be perfectly safe, we have to write an unmanaged wrapper in C/C++.  But I don't see why it would add such a call between two native calls using simple (not managed objects) values, thus I think this is "safe enough" for what it is used for.

Reply:
Ok, sounds good!